appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java

```diff
@@ -2295,6 +2332,155 @@ private boolean runAaptPackage(File manifestFile, File resDir, String tmpPackage
+  }
+
+  private boolean runAapt2Link(File manifestFile, String tmpPackageName, File symbolOutputDir) {
+    String aapt2Tool;
+    String osName = System.getProperty("os.name");
+    if (osName.equals("Mac OS X")) {
+      aapt2Tool = MAC_AAPT2_TOOL;
+    } else if (osName.equals("Linux")) {
+      aapt2Tool = LINUX_AAPT2_TOOL;
+    } else if (osName.startsWith("Windows")) {
+      aapt2Tool = WINDOWS_AAPT2_TOOL;
+    } else {
+      LOG.warning("YAIL compiler - cannot run AAPT2 on OS " + osName);
+      err.println("YAIL compiler - cannot run AAPT2 on OS " + osName);
+      userErrors.print(String.format(ERROR_IN_STAGE, "AAPT2"));
+      return false;
+    }
+    appRTxt = new File(symbolOutputDir, "R.txt");
+
+    List<String> aapt2CommandLine = new ArrayList<>();
+    aapt2CommandLine.add(getResource(aapt2Tool));
+    aapt2CommandLine.add("link");
+    aapt2CommandLine.add("--proto-format");
+    aapt2CommandLine.add("-o");
+    aapt2CommandLine.add(tmpPackageName);
+    aapt2CommandLine.add("-I");
+    aapt2CommandLine.add(getResource(ANDROID_RUNTIME));
+    aapt2CommandLine.add("-R");
+    aapt2CommandLine.add(resourcesZip.getAbsolutePath());
+    aapt2CommandLine.add("-A");
+    aapt2CommandLine.add(createDir(project.getBuildDirectory(), ASSET_DIR_NAME).getAbsolutePath());
+    aapt2CommandLine.add("--manifest");
+    aapt2CommandLine.add(manifestFile.getAbsolutePath());
+    aapt2CommandLine.add("--output-text-symbols");
+    aapt2CommandLine.add(appRTxt.getAbsolutePath());
+    aapt2CommandLine.add("--auto-add-overlay");
+    aapt2CommandLine.add("--no-version-vectors");
+    aapt2CommandLine.add("--no-auto-version");
+    aapt2CommandLine.add("--no-version-transitions");
+    aapt2CommandLine.add("--no-resource-deduping");
+    aapt2CommandLine.add("-v");
+    String[] aapt2LinkCommandLine = aapt2CommandLine.toArray(new String[0]);
+
+    long startAapt2 = System.currentTimeMillis();
+    if (!Execution.execute(null, aapt2LinkCommandLine, System.out, System.err)) {
+      LOG.warning("YAIL compiler - AAPT2 link execution failed.");
+      err.println("YAIL compiler - AAPT2 link execution failed.");
+      userErrors.print(String.format(ERROR_IN_STAGE, "AAPT2 link"));
+      return false;
+    }
+
+    String aaptTimeMessage = "AAPT2 link time: " + ((System.currentTimeMillis() - startAapt2) / 1000.0) + " seconds";
+    out.println(aaptTimeMessage);
+    LOG.info(aaptTimeMessage);
+    return true;
+  }
+
+  private boolean bundleTool(File buildDir, int childProcessRam, String tmpPackageName,
+                             String outputFileName, File deployDir, String keystoreFilePath, String dexedClassesDir) {
+    try {
+      String osName = System.getProperty("os.name");
+      String jarsignerTool;
+      if (osName.equals("Mac OS X")) {
+        jarsignerTool = System.getenv("JAVA_HOME") + "/bin/jarsigner";
+      } else if (osName.equals("Linux")) {
+        jarsignerTool = System.getenv("JAVA_HOME") + "/bin/jarsigner";
```

Comment:
This doesn't work. JAVA_HOME isn't necessarily set in the buildserver environment. You should use System.getProperty("java.home"). However this gives you the JRE home. So you have to do a ../bin/jarsigner to get to it. You can construct the path identically for Mac OS and Linux. But you have to separate out Windows because of the need to add the .exe on the end of the jarsigner binary name.

Reply:
If we use the JDK in the Docker image, we can bypass all this by invoking jarsigner without any qualification as it'll be in the PATH. I've tried this on macOS 10.15, Windows 10, and Ubuntu 18.04 and it worked (no need to even specify .exe on Windows). However, since we only use the JRE in the buildserver image, we would need to embed the jarsigner executable similarly to other tools (aapt, apksigner, etc.). Since we don't have the rights to redistribute jarsigner, it seems like using the JDK in the buildserver is the safer route.

Reply:
We already embed jarsigner (the buildserver has the jre and a few additions from the jdk). We used to use it a few versions ago before we switched to apksigner.